<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML321ENSkillsNetwork817-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Content-based Course Recommender System using Course Similarities**


Estimated time needed: **45** minutes


In one of the previous lab, you have learned and practiced how to calculate the similarity between two courses using Bag of Words (BoW) features. For example, the similarity between course1 `Machine Learning for Everyone` and course2 `Machine Learning for Beginners` are `75%` as shown below.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-ML321EN-SkillsNetwork/labs/module_2/images/course_sim.png)


As we mentioned before, the content-based recommender system is highly based on the similarity calculation among items. The similarity or closeness of items is measured based on the similarity in the content or features of those items. The course genres are important features, and in addition to that, the BoW value is another important type of feature to represent course textual content. 


In this lab, you will apply the course similarities metric to recommend  new courses which are similar to a user's presently enrolled courses.


## Objectives


After completing this lab you will be able to:


* Obtain the similarity between courses from a course similarity matrix
* Use the course similarity matrix to find and recommend new courses which are similar to enrolled courses


----


## Prepare and setup lab environment


Let's first install and import the required libraries:


In [14]:
!pip install seaborn
!pip install numpy
!pip install pandas
!pip install matplotlib


[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip


In [15]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

%matplotlib inline

In [16]:
# also set a random state
rs = 123

Next, let's load a pre-made course similarity matrix. If you are interested, you could easily calculate such a similarity matrix by iterating through all possible course pairs and calculating their similarities.


In [18]:
sim_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-ML321EN-SkillsNetwork/labs/datasets/sim.csv"

In [19]:
sim_df = pd.read_csv(sim_url)
sim_df

0         1         2         3         4         5         6  \
0    1.000000  0.088889  0.088475  0.065556  0.048810  0.104685  0.065202   
1    0.088889  1.000000  0.055202  0.057264  0.012182  0.078379  0.032545   
2    0.088475  0.055202  1.000000  0.026463  0.039406  0.000000  0.000000   
3    0.065556  0.057264  0.026463  1.000000  0.000000  0.250490  0.390038   
4    0.048810  0.012182  0.039406  0.000000  1.000000  0.000000  0.000000   
..        ...       ...       ...       ...       ...       ...       ...   
302  0.033944  0.028239  0.018270  0.094759  0.060474  0.064851  0.053856   
303  0.076825  0.063911  0.082698  0.030638  0.030415  0.000000  0.000000   
304  0.072898  0.138270  0.133400  0.017443  0.129871  0.009285  0.000000   
305  0.039276  0.031367  0.012684  0.018796  0.000000  0.015008  0.024926   
306  0.121113  0.076940  0.000000  0.158073  0.000000  0.126211  0.157219   

            7         8         9  ...       297       298       299  \
0    0.143346  0.000000  0.024405  ...  0.012695  0.070225  0.058224   
1    0.119251  0.044162  0.000000  ...  0.180593  0.124631  0.087187   
2    0.154303  0.000000  0.000000  ...  0.040996  0.037796  0.013430   
3    0.000000  0.000000  0.000000  ...  0.151882  0.420084  0.427908   
4    0.085126  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
..        ...       ...       ...  ...       ...       ...       ...   
302  0.039467  0.036539  0.060474  ...  0.047186  0.116008  0.137406   
303  0.119098  0.055132  0.045622  ...  0.177989  0.092381  0.145126   
304  0.254274  0.094165  0.025974  ...  0.153128  0.105191  0.138692   
305  0.082199  0.076102  0.055978  ...  0.149234  0.093962  0.162170   
306  0.000000  0.106668  0.039231  ...  0.142850  0.250859  0.260731   

          300       301       302       303       304       305       306  
0    0.046610  0.025850  0.033944  0.076825  0.072898  0.039276  0.121113  
1    0.093060  0.019354  0.028239  0.063911  0.138270  0.031367  0.076940  
2    0.037630  0.000000  0.018270  0.082698  0.133400  0.012684  0.000000  
3    0.055764  0.000000  0.094759  0.030638  0.017443  0.018796  0.158073  
4    0.096877  0.000000  0.060474  0.030415  0.129871  0.000000  0.000000  
..        ...       ...       ...       ...       ...       ...       ...  
302  0.211743  0.256220  1.000000  0.211520  0.168595  0.129766  0.263734  
303  0.527636  0.249675  0.211520  1.000000  0.242269  0.416067  0.178384  
304  0.281108  0.132977  0.168595  0.242269  1.000000  0.220159  0.128902  
305  0.476644  0.222350  0.129766  0.416067  0.220159  1.000000  0.126274  
306  0.099901  0.249325  0.263734  0.178384  0.128902  0.126274  1.000000  

[307 rows x 307 columns]

The similarity matrix is a real number, symmetric metric with each element representing the similarity value (ranged 0 to 1) between course index `i` and course index `j`. 


We could use `seaborn` to visualize the similarity metric, and since it is symmetric, we can just show the triangular  matrix (lower left):


In [22]:
# Configure seaborn to set the plot style to 'white'
sns.set_theme(style="white")

# Create a mask for the upper triangle of the similarity matrix
mask = np.triu(np.ones_like(sim_df, dtype=bool))

# Create a new figure and axis for the heatmap
_, ax = plt.subplots(figsize=(11, 9))

# Create a diverging color palette for the heatmap
cmap = sns.diverging_palette(230, 20, as_cmap=True)

# Plot a similarity heat map using seaborn's heatmap function
sns.heatmap(sim_df, mask=mask, cmap=cmap, vmin=0.01, vmax=1, center=0,
            square=True)


<Axes: >

As we can see from the heatmap; there are many hot spots, which means many courses are similar to each other. Such patterns suggest that it is possible to build a recommender system based on course similarities. 


Let's take a look at a quick example:


In [25]:
# Let's first load the course content and BoW dataset
course_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-ML321EN-SkillsNetwork/labs/datasets/course_processed.csv"
course_df = pd.read_csv(course_url)
bow_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-ML321EN-SkillsNetwork/labs/datasets/courses_bows.csv"
bow_df = pd.read_csv(bow_url)

In [26]:
bow_df.head()

doc_index    doc_id   token  bow
0          0  ML0201EN      ai    2
1          0  ML0201EN    apps    2
2          0  ML0201EN   build    2
3          0  ML0201EN   cloud    1
4          0  ML0201EN  coming    1

First, we want to mention that the matrix indices are course indices (such as `0, 1, 2, 3`). Very often we need to query the actual course ids (such as `ML0151EN` and `ML0101ENv3`) based on course indices and vice versa. We can save the course id's and indices into two dictionaries for late queries:


Then, based on the `doc_index` and `doc_id` columns, we create an index to id mapping and another id to index mapping in two Python dictionaries:


In [29]:
# Create course id to index and index to id mappings
def get_doc_dicts(bow_df):
    # Group the DataFrame by course index and ID, and get the maximum value for each group
    grouped_df = bow_df.groupby(['doc_index', 'doc_id']).max().reset_index(drop=False)
    # Create a dictionary mapping indices to course IDs
    idx_id_dict = grouped_df[['doc_id']].to_dict()['doc_id']
    # Create a dictionary mapping course IDs to indices
    id_idx_dict = {v: k for k, v in idx_id_dict.items()}
    # Clean up temporary DataFrame
    del grouped_df
    return idx_id_dict, id_idx_dict

Now suppose we have two example courses:


In [31]:
course1 = course_df[course_df['COURSE_ID'] == "ML0151EN"]
course1

COURSE_ID                    TITLE  \
200  ML0151EN  machine learning with r   

                                           DESCRIPTION  
200  this machine learning with r course dives into...

In [32]:
course2 = course_df[course_df['COURSE_ID'] == "ML0101ENv3"]
course2

COURSE_ID                         TITLE  \
158  ML0101ENv3  machine learning with python   

                                           DESCRIPTION  
158  machine learning can be an incredibly benefici...

From their titles we can see they are all about machine learning. As such, they should be very similar to each other. Let's try to find their similarity in the similarity matrix.

With their course ids, we can use the `id_idx_dict` dictionary to query their row and column index on the similarity matrix:


In [34]:
idx_id_dict, id_idx_dict = get_doc_dicts(bow_df)
idx1 = id_idx_dict["ML0151EN"]
idx2 = id_idx_dict["ML0101ENv3"]
print(f"Course 1's index is {idx1} and Course 2's index is {idx2}")

Course 1's index is 200 and Course 2's index is 158


Then we can locate their similarity value in row 200 and col 158, `sim_matrix[200][158]`:


In [36]:
sim_matrix = sim_df.to_numpy()

In [37]:
sim = sim_matrix[idx1][idx2]
sim

0.6626221399549089

It's about 66% meaning these two courses are quite similar to each other.


### TASK: Find courses which are similar enough to your enrolled courses.


Now you know how to easily use the pre-computed similarity matrix to query the similarity between any two courses. Do you want to make some course recommendations for yourself?

Let's assume you are an end-user of the online course platform and already audited or completed some courses previously. Next, you expect the system would recommend similar courses based on your enrollments history.


From the full course list, choose any courses that may interest you, such as those machine learning related courses:


In [42]:
pd.set_option('display.max_rows', None)
pd.set_option('max_colwidth', None)
course_df[['COURSE_ID', 'TITLE']]

COURSE_ID  \
0      ML0201EN   
1      ML0122EN   
2    GPXX0ZG0EN   
3      RP0105EN   
4    GPXX0Z2PEN   
5      CNSC02EN   
6      DX0106EN   
7    GPXX0FTCEN   
8    RAVSCTEST1   
9    GPXX06RFEN   
10   GPXX0SDXEN   
11     CC0271EN   
12     WA0103EN   
13     DX0108EN   
14   GPXX0PICEN   
15     DAI101EN   
16   GPXX0W7KEN   
17   GPXX0QR3EN   
18     BD0145EN   
19     HCC105EN   
20     DE0205EN   
21     DS0132EN   
22     OS0101EN   
23     DS0201EN   
24     BENTEST4   
25     CC0210EN   
26     PA0103EN   
27     HCC104EN   
28   GPXX0A1YEN   
29    TMP0105EN   
30     PA0107EN   
31     DB0113EN   
32     PA0109EN   
33    PHPM002EN   
34   GPXX03HFEN   
35       RP0103   
36     RP0103EN   
37     BD0212EN   
38    GPXX0IBEN   
39     SECM03EN   
40     SC0103EN   
41   GPXX0YXHEN   
42     RP0151EN   
43       TA0105   
44     SW0201EN   
45      TMP0106   
46   GPXX0BUBEN   
47     ST0201EN   
48     ST0301EN   
49     SW0101EN   
50    TMP0101EN   
51     DW0101EN   
52     BD0143EN   
53     WA0101EN   
54   GPXX04HEEN   
55     BD0141EN   
56     CO0401EN   
57   ML0122ENv1   
58     BD0151EN   
59     TA0106EN   
60       TMP107   
61     ML0111EN   
62   GPXX048OEN   
63     CO0201EN   
64   GPXX01DCEN   
65   GPXX04XJEN   
66   GPXX0JZ4EN   
67   GPXX0ZYVEN   
68   GPXX0ZMZEN   
69   GPXX0742EN   
70   GPXX0KV4EN   
71   GPXX01RYEN   
72     CC0120EN   
73     QC0101EN   
74   GPXX0YMEEN   
75   GPXX0Q8AEN   
76     TA0105EN   
77   GPXX0XFQEN   
78    GPXX07REN   
79     PA0101EN   
80     BC0202EN   
81     BC0101EN   
82   GPXX0725EN   
83     BC0201EN   
84   GPXX0MIIEN   
85     BD0223EN   
86   GPXX0435EN   
87   GPXX06ZLEN   
88   GPXX0QS6EN   
89   GPXX07YGEN   
90     AI0111EN   
91   GPXX0QJFEN   
92   GPXX0LLEEN   
93     BD0121EN   
94     BD0135EN   
95     BD0133EN   
96     BD0131EN   
97     BD0115EN   
98   GPXX04V3EN   
99   GPXX05P1EN   
100  GPXX0D14EN   
101  GPXX0G3KEN   
102    BD0211EN   
103  GPXX0HZ2EN   
104  GPXX04TNEN   
105  GPXX0IHMEN   
106    ST0101EN   
107    BD0153EN   
108    DP0101EN   
109    BD0137EN   
110    ML0120EN   
111  GPXX0E3QEN   
112  GPXX0XV3EN   
113    CB0201EN   
114  CB0105ENv1   
115  GPXX0NHZEN   
116    CL0101EN   
117    CO0301EN   
118  GPXX0XENEN   
119  GPXX0HAAEN   
120    DA0201EN   
121    DJ0101EN   
122    CP0101EN   
123    DB0111EN   
124    DB0115EN   
125    COM001EN   
126    DA0151EN   
127    DB0151EN   
128    CO0302EN   
129  GPXX0T0FEN   
130      DS0107   
131    DS0110EN   
132    DX0107EN   
133    DS0321EN   
134    DS0105EN   
135    DS0103EN   
136  GPXX0I4FEN   
137  GPXX06KEEN   
138  GPXX0YBFEN   
139  LB0109ENv1   
140  GPXX0KHHEN   
141    LB0111EN   
142    ML0115EN   
143  LB0105ENv1   
144  LB0103ENv1   
145    IT0101EN   
146    CC0250EN   
147    BD0101EN   
148    EE0101EN   
149  GPXX04P5EN   
150    CO0101EN   
151  ML0122ENv3   
152  LB0107ENv1   
153    CB0101EN   
154  GPXX0PG8EN   
155    ML0101EN   
156    CB0103EN   
157    ML0109EN   
158  ML0101ENv3   
159  GPXX0M7ZEN   
160    CC0121EN   
161    DA0101EN   
162  GPXX07UGEN   
163  GPXX0QU9EN   
164    SC0101EN   
165  GPXX0QTEEN   
166  GPXX0WTIEN   
167  GPXX08WYEN   
168  GPXX0UMSEN   
169  GPXX0M6UEN   
170  GPXX097UEN   
171    SN0111EN   
172  GPXX0JGFEN   
173    CO0193EN   
174  GPXX0HC7EN   
175  GPXX04MXEN   
176    DS0101EN   
177  LB0101ENv1   
178    DS0301EN   
179  GPXX0RL8EN   
180  GPXX05LMEN   
181    BD0111EN   
182    CC0101EN   
183  GPXX0UN5EN   
184    BD0221EN   
185    CC0103EN   
186  GPXX0TY1EN   
187  GPXX0T3CEN   
188    PY0101EN   
189    CC0201EN   
190    BD0123EN   
191    DB0101EN   
192  GPXX05RDEN   
193    DV0151EN   
194  GPXX0BSAEN   
195  GPXX0G81EN   
196  GPXX0FFCEN   
197  GPXX0MP0EN   
198    DV0101EN   
199    ML0103EN   
200    ML0151EN   
201  GPXX01AVEN   
202  ML0120ENv2   
203    RP0101EN   
204  GPXX0G31EN   
205  GPXX0KY1EN   
206  GPXX0JLHEN   
207    CC0150EN   
208  GPXX0QQ3EN   
209  GPXX0RQLEN 

In [43]:
# Reset pandas settings
pd.reset_option('display.max_rows')
pd.reset_option('max_colwidth')

_TODO: Browse the course list and choose your interested courses_ 


In [45]:
enrolled_course_ids = ["ML0151EN", "excourse47"] # add your interested coures id to the list

In [46]:
enrolled_courses = course_df[course_df['COURSE_ID'].isin(enrolled_course_ids)]
enrolled_courses

COURSE_ID                     TITLE  \
200    ML0151EN   machine learning with r   
260  excourse47  machine learning for all   

                                           DESCRIPTION  
200  this machine learning with r course dives into...  
260  machine learning  often called artificial inte...

Given the full course list, we can find those unselected courses:


In [48]:
all_courses = set(course_df['COURSE_ID'])

In [49]:
unselected_course_ids = all_courses.difference(enrolled_course_ids)
unselected_course_ids

{'AI0111EN',
 'BC0101EN',
 'BC0201EN',
 'BC0202EN',
 'BD0101EN',
 'BD0111EN',
 'BD0115EN',
 'BD0121EN',
 'BD0123EN',
 'BD0131EN',
 'BD0133EN',
 'BD0135EN',
 'BD0137EN',
 'BD0141EN',
 'BD0143EN',
 'BD0145EN',
 'BD0151EN',
 'BD0153EN',
 'BD0211EN',
 'BD0212EN',
 'BD0221EN',
 'BD0223EN',
 'BENTEST4',
 'CB0101EN',
 'CB0103EN',
 'CB0105ENv1',
 'CB0201EN',
 'CC0101EN',
 'CC0103EN',
 'CC0120EN',
 'CC0121EN',
 'CC0150EN',
 'CC0201EN',
 'CC0210EN',
 'CC0250EN',
 'CC0271EN',
 'CL0101EN',
 'CNSC02EN',
 'CO0101EN',
 'CO0193EN',
 'CO0201EN',
 'CO0301EN',
 'CO0302EN',
 'CO0401EN',
 'COM001EN',
 'CP0101EN',
 'DA0101EN',
 'DA0151EN',
 'DA0201EN',
 'DAI101EN',
 'DB0101EN',
 'DB0111EN',
 'DB0113EN',
 'DB0115EN',
 'DB0151EN',
 'DE0205EN',
 'DJ0101EN',
 'DP0101EN',
 'DS0101EN',
 'DS0103EN',
 'DS0105EN',
 'DS0107',
 'DS0110EN',
 'DS0132EN',
 'DS0201EN',
 'DS0301EN',
 'DS0321EN',
 'DV0101EN',
 'DV0151EN',
 'DW0101EN',
 'DX0106EN',
 'DX0107EN',
 'DX0108EN',
 'EE0101EN',
 'GPXX01AVEN',
 'GPXX01DCEN',
 'GPXX01

Now, you can iterate each unselect course and check if it is similar enough to any of your selected courses. If the similarity is larger than a threshold such as 0.5 or 0.6, then add it to your course recommendation list:


_TODO: Complete the following method to recommend courses which are similar to your enrolled courses_


In [52]:
def generate_recommendations_for_one_user(enrolled_course_ids, unselected_course_ids, id_idx_dict, sim_matrix):
    # Create a dictionary to store your recommendation results
    res = {}
    # Set a threshold for similarity
    threshold = 0.6
    # Iterate over enrolled courses
    for enrolled_course in enrolled_course_ids:
        # Iterate over unselected courses
        for unselect_course in unselected_course_ids:
            # Check if both enrolled and unselected courses exist in the id_idx_dict
            if enrolled_course in id_idx_dict and unselect_course in id_idx_dict:
                # Initialize similarity value
                sim = 0
                # Find the two indices for each enrolled_course and unselect_course, based on their two ids
                # Calculate the similarity between an enrolled_course and an unselect_course
                # e.g., Course ML0151EN's index is 200 and Course ML0101ENv3's index is 158
                
                idx1 = id_idx_dict[enrolled_course]
                idx2 = id_idx_dict[unselect_course]

                
                # Find the similarity value from the sim_matrix
                # sim = sim_matrix[200][158]

                sim = sim_matrix[idx1][idx2]
                
                 # Check if the similarity exceeds the threshold
                if sim > threshold:
                    # Update recommendation dictionary with course ID and similarity score
                    if unselect_course not in res:
                        # If the unselected course is not already in the recommendation dictionary (`res`), add it.
                        res[unselect_course] = sim
                    else:
                        # If the unselected course is already in the recommendation dictionary (`res`), compare the similarity score.
                        # If the current similarity score is greater than or equal to the existing similarity score for the course,
                        # update the similarity score in the recommendation dictionary (`res`) with the current similarity score.
                        if sim >= res[unselect_course]:
                            res[unselect_course] = sim
                            
    # Sort the results by similarity
    res = {k: v for k, v in sorted(res.items(), key=lambda item: item[1], reverse=True)}
     # Return the recommendation dictionary
    return res


generate_recommendations_for_one_user(enrolled_course_ids, unselected_course_ids, id_idx_dict, sim_matrix)

{'excourse46': 0.6892532539898828,
 'ML0101ENv3': 0.6626221399549089,
 'excourse60': 0.6307666542363487}

<details>
    <summary>Click here for Hints</summary>
    
You can find the indices of the courses by using `id_idx_dict` dictionary, for example `id_idx_dict[enrolled_course]`. Then use `sim_matrix` to find the similarity of the courses as was shown earlier in the lab. 
    


The completed `generate_recommendations_user(...)` may ouput a dictionary like this:


{'ML0151EN': 0.6626221399549089,
 'excourse47': 0.6347547807096177,
 'excourse46': 0.6120541193300345}


### TASK: Generate course recommendations based on course similarities for all test uesrs


In the previous task, you made some recommendations for yourself. Next, let's try to make recommendations for all the test users in the test dataset.


In [58]:
test_users_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMSkillsNetwork-ML0321EN-Coursera/labs/v2/module_3/ratings.csv"
test_users_df = pd.read_csv(test_users_url)

Let's look at how many test users we have in the dataset.


In [60]:
test_users = test_users_df.groupby(['user']).max().reset_index(drop=False)
test_user_ids = test_users['user'].to_list()
print(f"Total numbers of test users {len(test_user_ids)}")

Total numbers of test users 33901


_TODO: Complete the generate_recommendation_scores() function blow to generate recommendation score for all users. You may also implement the task with different solutions._

In [62]:
# WRITE YOUR CODE HERE
def generate_recommendations_for_all():
    users = []
    courses = []
    sim_scores = []
    # Test user dataframe
    # Course similarity matrix
    sim_df = pd.read_csv(sim_url)
    sim_matrix = sim_df.to_numpy()
    # Course content dataframe
    course_df = pd.read_csv(course_url)
    # Course BoW features
    bow_df = pd.read_csv(bow_url)
    test_users = test_users_df.groupby(['user']).max().reset_index(drop=False)
    test_user_ids = test_users['user'].to_list()

    # ...

    for user_id in test_user_ids:
        idx_id_dict, id_idx_dict = get_doc_dicts(bow_df)

        enrolled_course_ids = test_users_df[test_users_df['user'] == 1502801]['item'].to_numpy()
        all_courses = set(course_df['COURSE_ID'])
        unselected_course_ids = all_courses.difference(enrolled_course_ids)

        course_score_dict = generate_recommendations_for_one_user(enrolled_course_ids, unselected_course_ids, id_idx_dict, sim_matrix)
        # For each user, call generate_recommendations_for_one_user() to generate the recommendation results
        # Save the result to courses, sim_scores list
        for k, v in course_score_dict.items():
            users.append(user_id)
            courses.append(k)
            sim_scores.append(v)

    return users, courses, sim_scores

<details>
    <summary>Click here for Hints</summary>
    
Note that you can use `generate_recommendations_for_one_user` function to find the list of recommended courses for each user. Find the `enrolled_course_ids` list by simply finding them in the `test_users` list by running `test_users[test_users['user']==user_id]['item']` and to find the unselected courses list you can apply `all_courses.difference()` with `enrolled_course_ids` as it's parameter (as done earlier in the lab). (Keep the last 2 parameters of `generate_recommendations_for_one_user` the same)
</details>


After you completed the `generate_recommendations_for_all()` function, you can call it to save the results into a dataframe:


In [70]:
res_dict = {}
users, courses, sim_scores = generate_recommendations_for_all()
res_dict['USER'] = users
res_dict['COURSE_ID'] = courses
res_dict['SCORE'] = sim_scores
res_df = pd.DataFrame(res_dict, columns=['USER', 'COURSE_ID', 'SCORE'])

In [71]:
res_df.head()

USER   COURSE_ID     SCORE
0     2  excourse67  0.708214
1     2  excourse72  0.652535
2     2  excourse74  0.650071
3     2    BD0145EN  0.623544
4     2  excourse68  0.616759

Similar to the previous user profile and course genre lab, with the recommendations generated for each user, you need to write some extra analytic code to answer the following questions:


- On average, how many new/unseen courses have been recommended to each user?
- What are the most frequently recommended courses? Return the top-10 commonly recommended courses across all users?


For example, suppose we have only 3 test users, each user receives the following recommendations:


- User1: ['course1', 'course2']
- User2: ['course3', 'course4']
- User3: ['course3', 'course4', 'course5']


Then, the average recommended courses per user is $(2 + 2 + 3) / 3 = 2.33$. The top-2 recommended courses are: `course3`: 2 times, and `course4`: 2 times.


Note that the answers may depend on your similarity threshold (default is 0.6). A lower similarity threshold yields more recommended courses but with smaller irrelevance.

Ideally, we should limit the maximum course recommendations for each user to be less than 20 courses per user.


## Authors


[Yan Luo]((https://www.linkedin.com/in/yan-luo-96288783/)


### Other Contributors


```toggle## Change Log
```


```toggle|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
```
```toggle|-|-|-|-|
```
```toggle|2021-10-25|1.0|Yan|Created the initial version|
```


Copyright © 2021 IBM Corporation. All rights reserved.
